platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java

```diff
@@ -112,99 +122,113 @@ private File getJavaHome(File markedLocation) {
      */
-    public File provisionFromArchive(JavaToolchainSpec spec, File jdkArchive, URI uri) {
-        final File[] unpackFolder = new File[1];
-        final File[] installFolder = new File[1];
+    public File provisionFromArchive(JavaToolchainSpec spec, File jdkArchive, URI uri) throws IOException {
+        // Unpack into temporary directory (but on same file system as our target directory location)
+        File unpackFolder = unpack(jdkArchive);
         try {
-            unpackFolder[0] = unpack(jdkArchive);
-
-            //put the marker file in the unpack folder from where it will get in its proper place
-            //when the contents are copied to the installation folder
-            File markedLocation = markLocationInsideFolder(unpackFolder[0]);
+            // Get the folder that is the real root of the unpacked JDK, skipping any archive root folder
+            File unpackedRoot = determineUnpackedRoot(unpackFolder);
 
-            //probe unpacked installation for its metadata
-            JvmInstallationMetadata metadata = getMetadata(markedLocation);
+            // Probe unpacked installation for its metadata
+            JvmInstallationMetadata metadata = getMetadata(unpackedRoot);
 
             validateMetadataMatchesSpec(spec, uri, metadata);
 
-            String installFolderName = getInstallFolderName(metadata);
-            installFolder[0] = new File(jdkDirectory, installFolderName);
-
-            //make sure the install folder is empty
-            checkInstallFolderForLeftoverContent(installFolder[0], uri, spec, metadata);
-            operations.delete(installFolder[0]);
-
-            //copy content of unpack folder to install folder, including the marker file
-            operations.copy(copySpec -> {
-                copySpec.from(unpackFolder[0]);
-                copySpec.into(installFolder[0]);
-            });
-
-            LOGGER.info("Installed toolchain from {} into {}", uri, installFolder[0]);
-            return getJavaHome(markedLocation(installFolder[0]));
-        } catch (Throwable t) {
-            // provisioning failed, clean up leftovers
-            if (installFolder[0] != null) {
-                operations.delete(installFolder[0]);
+            // Check our target directory, to see if anything exists there, and if so, is it marked as ready?
+            File installFolder = new File(jdkDirectory, getInstallFolderName(metadata));
+            if (!installFolder.getParentFile().mkdirs() && !installFolder.getParentFile().isDirectory()) {
+                throw new IOException("Failed to create install parent directory: " + installFolder.getParentFile());
             }
-            throw t;
+            // Before checking existence, lock the install folder name to prevent concurrent installations
+            try (FileLock ignored = acquireWriteLock(installFolder, "Provisioning JDK from " + uri)) {
```

Comment:
❓ It looks like this write lock is already acquired in `DefaultJavaToolchainProvisioningService#provisionInstallation`. It seems like we shouldn't need to acquire it again here. 

While this lock is on the downloaded archive, I would expect that the archive file lock would be sufficient

Reply:
This is to catch a rare case where there are two different archives that contain the same JVM metadata-wise. Two archives can point to the same install folder. Perhaps though, with this check, it could be possible to drop the outer lock. But that lock does guard against doing an expensive unpack of the archive, so I think it's worth it.

Reply:
Yeah I think we should definitely be locking on the archive. Don't want download that in parallel.